provisioning/metrics_service.go

```diff
@@ -0,0 +1,135 @@
+package provisioning
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+)
+
+const (
+	metricsService          = "metal3-metrics"
+	metal3ExposeMetricsPort = 8445
+	// Port where BMO metrics can be found
+	defaultMetal3MetricsAddress = ":60000"
+	certStoreName               = "cluster-baremetal-operator-tls"
+)
+
+// This Service Monitor should be able to discover the metal3-metrics
+// NodePort Service. To do that it looks for Services that satisfy
+// the matchLabels and the named port specified in the endpoints below.
+
+const metal3ServiceMonitorDefinition = `
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  namespace: openshift-machine-api
+  name: metal3-metrics-servicemonitor
+  labels:
+    cboLabelName: metricsService
+  annotations:
+    exclude.release.openshift.io/internal-openshift-hosted: "true"
+spec:
+  namespaceSelector:
+    matchNames:
+      - openshift-machine-api
+  selector:
+    matchLabels:
+      cboLabelName: metricsService
+ endpoints:
+  - port: metal3-mtrc
+    bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+    interval: 30s
+    scheme: https
+    tlsConfig:
+      caFile: /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt
+      serverName: cluster-baremetal-operator.openshift-machine-api.svc
+`
+
+// This NodePort Service would is created to expose metrics from pods in the
+// metal3 deployment. The Selector is used to match Pods that this Service
+// represents. Since the metal3 deployment pods could contain the 4.6 or 4.7
+// based labels, the Selector should also be constructed that way.
+// The Named port in the ServiceMonitor is defined as a ServicePort here.
```

Comment:
this comment above is not very understandable

Reply:
Agree. I have updated it. Hopefully, it is better.